### DISCUSSION TITLE
Auto-increment primary key

### DISCUSSION THREAD

**Discussion by 3620adnanserki** [jorenvanhee/joren.co#3] [General]
hello! I loved the blog as it helped me understand macro and use it in my code.  I had a general observation when record is Upsert then auto incremented primary key jumps to the number of records in the data. I didn't quite understood that, even after i removed DB transaction.

**Comment by jorenvanhee** [comment]
Looks like this is related to how MySQL handles upserts. Found [this](https://github.com/laravel/framework/discussions/38938) after a quick google. Wasn't aware of that, thanks for letting me know!

I guess you could always replace `upsert` by looping over your records and using `updateOrCreate`.

**Reply by 3620adnanserki** [reply]
Sorry for the delayed response. Thats exactly what I did. Here is what I came up with:

```
protected function updateOrCreateRecords()
    {
        $values = $this->records->toArray();
        foreach ($values as $value) {
            $recordKeyName = $this->query->getRelated()->getKeyName();

            $value[$recordKeyName] = array_key_exists($recordKeyName, $value)
                ? $value[$recordKeyName]
                : null;


            (clone $this->query)->updateOrCreate(
                [$recordKeyName => $value[$recordKeyName]],
                $value
            );
        }
    }
```

**Reply by jorenvanhee** [reply]
Thanks for sharing 👍
